src/read.rs

```diff
@@ -20,27 +20,21 @@ impl<'a, I> ReadHtml<'a, I> where I: Iterator<Item=Event<'a>> {
 }
 
 impl<'a, I> io::Read for ReadHtml<'a, I> where I: Iterator<Item=Event<'a>> {
-    fn read(&mut self, mut buf: &mut [u8]) -> io::Result<usize> {
-        while self.current.len() < buf.len() {
-            if let Some(next) = self.events.next() {
-                try!(write!(&mut self.current, "{}", next));
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        // TODO handle buf.len() == 0
+        if self.current.len() == 0 {
+            if let Some(event) = self.events.next() {
+                // This feels like an unnecessary redirection. Can we do better?
+                // Write directly to buf and left over to self.current?
+                try!(write!(&mut self.current, "{}", event));
             } else {
-                break;
+                return Ok(0)
             }
         }
-
-        if self.current.len() < buf.len() {
-            let len = self.current.len();
-            buf[..len].clone_from_slice(&self.current[..]);
-            self.current.clear();
-            Ok(len)
-        } else {
-            let leftover = self.current.split_off(buf.len());
-            buf.clone_from_slice(&self.current[..]);
-            self.current.clear();
-            self.current.extend_from_slice(&leftover[..]);
-            Ok(buf.len())
-        }
+        let len = ::std::cmp::min(self.current.len(), buf.len());
+        buf[..len].clone_from_slice(&self.current[..len]);
+        self.current = self.current.split_off(len);
```

Comment:
Only way to do this efficiently I can see is http://is.gd/V9EkFm, but I'd really like to avoid unsafe code in this library.